File: Dormitory_Manager/lib/models/inspection_settings.dart
import 'package:intl/intl.dart';

/// 점호 설정 모델
/// ✅ 수정: inspectionDate 필드 추가 (특정 날짜에만 점호 가능)
class InspectionSettings {
  final int? id;
  final String settingName;
  final String startTime;
  final String endTime;
  final DateTime? inspectionDate;  // ✅ 신규: 점호 날짜
  final bool isEnabled;
  final bool cameraOnly;
  final bool exifValidationEnabled;
  final int exifTimeToleranceMinutes;
  final bool gpsValidationEnabled;
  final double? dormitoryLatitude;
  final double? dormitoryLongitude;
  final int? gpsRadiusMeters;
  final bool roomPhotoValidationEnabled;
  final String? applicableDays;
  final bool isDefault;
  final int? scheduleId;  // ✅ 신규: 연결된 캘린더 일정 ID
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final String? createdBy;

  InspectionSettings({
    this.id,
    required this.settingName,
    required this.startTime,
    required this.endTime,
    this.inspectionDate,
    this.isEnabled = true,
    this.cameraOnly = true,
    this.exifValidationEnabled = true,
    this.exifTimeToleranceMinutes = 10,
    this.gpsValidationEnabled = false,
    this.dormitoryLatitude,
    this.dormitoryLongitude,
    this.gpsRadiusMeters = 100,
    this.roomPhotoValidationEnabled = true,
    this.applicableDays = 'ALL',
    this.isDefault = false,
    this.scheduleId,
    this.createdAt,
    this.updatedAt,
    this.createdBy,
  });

  factory InspectionSettings.fromJson(Map<String, dynamic> json) {
    return InspectionSettings(
      id: json['id'],
      settingName: json['settingName'] ?? '',
      startTime: json['startTime'] ?? '21:00',
      endTime: json['endTime'] ?? '23:59',
      inspectionDate: json['inspectionDate'] != null
          ? DateTime.parse(json['inspectionDate'])
          : null,
      isEnabled: json['isEnabled'] ?? true,
      cameraOnly: json['cameraOnly'] ?? true,
      exifValidationEnabled: json['exifValidationEnabled'] ?? true,
      exifTimeToleranceMinutes: json['exifTimeToleranceMinutes'] ?? 10,
      gpsValidationEnabled: json['gpsValidationEnabled'] ?? false,
      dormitoryLatitude: json['dormitoryLatitude']?.toDouble(),
      dormitoryLongitude: json['dormitoryLongitude']?.toDouble(),
      gpsRadiusMeters: json['gpsRadiusMeters'] ?? 100,
      roomPhotoValidationEnabled: json['roomPhotoValidationEnabled'] ?? true,
      applicableDays: json['applicableDays'] ?? 'ALL',
      isDefault: json['isDefault'] ?? false,
      scheduleId: json['scheduleId'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      createdBy: json['createdBy'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'settingName': settingName,
      'startTime': startTime,
      'endTime': endTime,
      if (inspectionDate != null) 'inspectionDate': DateFormat('yyyy-MM-dd').format(inspectionDate!),
      'isEnabled': isEnabled,
      'cameraOnly': cameraOnly,
      'exifValidationEnabled': exifValidationEnabled,
      'exifTimeToleranceMinutes': exifTimeToleranceMinutes,
      'gpsValidationEnabled': gpsValidationEnabled,
      if (dormitoryLatitude != null) 'dormitoryLatitude': dormitoryLatitude,
      if (dormitoryLongitude != null) 'dormitoryLongitude': dormitoryLongitude,
      if (gpsRadiusMeters != null) 'gpsRadiusMeters': gpsRadiusMeters,
      'roomPhotoValidationEnabled': roomPhotoValidationEnabled,
      if (applicableDays != null) 'applicableDays': applicableDays,
      'isDefault': isDefault,
    };
  }

  /// ✅ 신규: 점호 날짜 포맷팅
  String? get formattedInspectionDate {
    if (inspectionDate == null) return null;
    return DateFormat('yyyy년 M월 d일').format(inspectionDate!);
  }

  /// ✅ 신규: 점호 날짜까지 남은 일수
  int? get daysUntilInspection {
    if (inspectionDate == null) return null;
    final today = DateTime.now();
    final todayOnly = DateTime(today.year, today.month, today.day);
    final inspectionOnly = DateTime(inspectionDate!.year, inspectionDate!.month, inspectionDate!.day);
    return inspectionOnly.difference(todayOnly).inDays;
  }

  /// ✅ 신규: 오늘이 점호 날짜인지 확인
  bool get isInspectionToday {
    if (inspectionDate == null) return true;  // 날짜 미설정시 매일 점호
    final today = DateTime.now();
    return inspectionDate!.year == today.year &&
        inspectionDate!.month == today.month &&
        inspectionDate!.day == today.day;
  }

  InspectionSettings copyWith({
    int? id,
    String? settingName,
    String? startTime,
    String? endTime,
    DateTime? inspectionDate,
    bool? isEnabled,
    bool? cameraOnly,
    bool? exifValidationEnabled,
    int? exifTimeToleranceMinutes,
    bool? gpsValidationEnabled,
    double? dormitoryLatitude,
    double? dormitoryLongitude,
    int? gpsRadiusMeters,
    bool? roomPhotoValidationEnabled,
    String? applicableDays,
    bool? isDefault,
    int? scheduleId,
  }) {
    return InspectionSettings(
      id: id ?? this.id,
      settingName: settingName ?? this.settingName,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      inspectionDate: inspectionDate ?? this.inspectionDate,
      isEnabled: isEnabled ?? this.isEnabled,
      cameraOnly: cameraOnly ?? this.cameraOnly,
      exifValidationEnabled: exifValidationEnabled ?? this.exifValidationEnabled,
      exifTimeToleranceMinutes: exifTimeToleranceMinutes ?? this.exifTimeToleranceMinutes,
      gpsValidationEnabled: gpsValidationEnabled ?? this.gpsValidationEnabled,
      dormitoryLatitude: dormitoryLatitude ?? this.dormitoryLatitude,
      dormitoryLongitude: dormitoryLongitude ?? this.dormitoryLongitude,
      gpsRadiusMeters: gpsRadiusMeters ?? this.gpsRadiusMeters,
      roomPhotoValidationEnabled: roomPhotoValidationEnabled ?? this.roomPhotoValidationEnabled,
      applicableDays: applicableDays ?? this.applicableDays,
      isDefault: isDefault ?? this.isDefault,
      scheduleId: scheduleId ?? this.scheduleId,
      createdAt: createdAt,
      updatedAt: updatedAt,
      createdBy: createdBy,
    );
  }
}

/// 점호 시간 확인 결과
/// ✅ 수정: 다음 점호 날짜 정보 추가
class InspectionTimeCheckResult {
  final bool allowed;
  final String message;
  final String? startTime;
  final String? endTime;
  final DateTime? nextInspectionDate;  // ✅ 신규: 다음 점호 날짜
  final int? daysUntilNext;            // ✅ 신규: 다음 점호까지 남은 일수

  InspectionTimeCheckResult({
    required this.allowed,
    required this.message,
    this.startTime,
    this.endTime,
    this.nextInspectionDate,
    this.daysUntilNext,
  });

  factory InspectionTimeCheckResult.fromJson(Map<String, dynamic> json) {
    return InspectionTimeCheckResult(
      allowed: json['allowed'] ?? false,
      message: json['message'] ?? '',
      startTime: json['startTime'],
      endTime: json['endTime'],
      nextInspectionDate: json['nextInspectionDate'] != null
          ? DateTime.parse(json['nextInspectionDate'])
          : null,
      daysUntilNext: json['daysUntilNext'],
    );
  }

  /// ✅ 신규: 다음 점호 날짜 포맷팅
  String? get formattedNextDate {
    if (nextInspectionDate == null) return null;
    return DateFormat('M월 d일').format(nextInspectionDate!);
  }
}

File: SpringBoot/src/main/java/com/dormitory/SpringBoot/domain/InspectionSettings.java
package com.dormitory.SpringBoot.domain;

import jakarta.persistence.*;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * 점호 설정 엔티티
 * ✅ 수정: inspectionDate 필드 추가 (특정 날짜에만 점호 가능하도록)
 */
@Entity
@Table(name = "inspection_settings")
public class InspectionSettings {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "setting_name", nullable = false, unique = true)
    private String settingName;

    @Column(name = "start_time", nullable = false)
    private LocalTime startTime;

    @Column(name = "end_time", nullable = false)
    private LocalTime endTime;

    // ✅ 신규 추가: 점호 날짜 (특정 날짜에만 점호 가능)
    @Column(name = "inspection_date")
    private LocalDate inspectionDate;

    @Column(name = "is_enabled", nullable = false)
    private Boolean isEnabled = true;

    @Column(name = "camera_only", nullable = false)
    private Boolean cameraOnly = true;

    @Column(name = "exif_validation_enabled", nullable = false)
    private Boolean exifValidationEnabled = true;

    @Column(name = "exif_time_tolerance_minutes", nullable = false)
    private Integer exifTimeToleranceMinutes = 10;

    @Column(name = "gps_validation_enabled", nullable = false)
    private Boolean gpsValidationEnabled = false;

    @Column(name = "dormitory_latitude")
    private Double dormitoryLatitude;

    @Column(name = "dormitory_longitude")
    private Double dormitoryLongitude;

    @Column(name = "gps_radius_meters")
    private Integer gpsRadiusMeters = 100;

    @Column(name = "room_photo_validation_enabled", nullable = false)
    private Boolean roomPhotoValidationEnabled = true;

    @Column(name = "applicable_days", length = 50)
    private String applicableDays = "ALL";

    @Column(name = "is_default", nullable = false)
    private Boolean isDefault = false;

    // ✅ 신규 추가: 캘린더 일정 ID (연동된 캘린더 일정)
    @Column(name = "schedule_id")
    private Long scheduleId;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "created_by", length = 50)
    private String createdBy;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    /**
     * ✅ 수정: 현재 날짜와 시간이 점호 허용 범위 내인지 확인
     */
    public boolean isWithinAllowedTime() {
        if (!Boolean.TRUE.equals(isEnabled)) {
            return false;
        }

        LocalDate today = LocalDate.now();
        LocalTime now = LocalTime.now();

        // ✅ 점호 날짜가 설정되어 있으면 해당 날짜인지 확인
        if (inspectionDate != null) {
            if (!today.equals(inspectionDate)) {
                return false;  // 오늘이 점호 날짜가 아니면 false
            }
        }

        // 시간 확인
        if (startTime.isBefore(endTime)) {
            return !now.isBefore(startTime) && !now.isAfter(endTime);
        } else {
            // 자정을 넘어가는 경우 (예: 23:00 ~ 01:00)
            return !now.isBefore(startTime) || !now.isAfter(endTime);
        }
    }

    /**
     * ✅ 신규: 오늘이 점호 날짜인지 확인
     */
    public boolean isInspectionDateToday() {
        if (inspectionDate == null) {
            return true;  // 날짜 미설정시 매일 점호 가능 (기존 방식)
        }
        return LocalDate.now().equals(inspectionDate);
    }

    /**
     * ✅ 신규: 점호 날짜까지 남은 일수 계산
     */
    public long getDaysUntilInspection() {
        if (inspectionDate == null) {
            return 0;
        }
        LocalDate today = LocalDate.now();
        return java.time.temporal.ChronoUnit.DAYS.between(today, inspectionDate);
    }

    /**
     * 현재 요일에 적용되는 설정인지 확인
     */
    public boolean isApplicableToday() {
        // ✅ 점호 날짜가 설정되어 있으면 요일 무시
        if (inspectionDate != null) {
            return LocalDate.now().equals(inspectionDate);
        }

        if (applicableDays == null || "ALL".equalsIgnoreCase(applicableDays)) {
            return true;
        }

        DayOfWeek today = LocalDateTime.now().getDayOfWeek();
        String todayStr = today.toString().substring(0, 3);

        Set<String> days = Arrays.stream(applicableDays.split(","))
                .map(String::trim)
                .map(String::toUpperCase)
                .collect(Collectors.toSet());

        return days.contains(todayStr);
    }

    // ==================== Getters and Setters ====================

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getSettingName() { return settingName; }
    public void setSettingName(String settingName) { this.settingName = settingName; }

    public LocalTime getStartTime() { return startTime; }
    public void setStartTime(LocalTime startTime) { this.startTime = startTime; }

    public LocalTime getEndTime() { return endTime; }
    public void setEndTime(LocalTime endTime) { this.endTime = endTime; }

    // ✅ 신규
    public LocalDate getInspectionDate() { return inspectionDate; }
    public void setInspectionDate(LocalDate inspectionDate) { this.inspectionDate = inspectionDate; }

    public Boolean getIsEnabled() { return isEnabled; }
    public void setIsEnabled(Boolean isEnabled) { this.isEnabled = isEnabled; }

    public Boolean getCameraOnly() { return cameraOnly; }
    public void setCameraOnly(Boolean cameraOnly) { this.cameraOnly = cameraOnly; }

    public Boolean getExifValidationEnabled() { return exifValidationEnabled; }
    public void setExifValidationEnabled(Boolean exifValidationEnabled) { this.exifValidationEnabled = exifValidationEnabled; }

    public Integer getExifTimeToleranceMinutes() { return exifTimeToleranceMinutes; }
    public void setExifTimeToleranceMinutes(Integer exifTimeToleranceMinutes) { this.exifTimeToleranceMinutes = exifTimeToleranceMinutes; }

    public Boolean getGpsValidationEnabled() { return gpsValidationEnabled; }
    public void setGpsValidationEnabled(Boolean gpsValidationEnabled) { this.gpsValidationEnabled = gpsValidationEnabled; }

    public Double getDormitoryLatitude() { return dormitoryLatitude; }
    public void setDormitoryLatitude(Double dormitoryLatitude) { this.dormitoryLatitude = dormitoryLatitude; }

    public Double getDormitoryLongitude() { return dormitoryLongitude; }
    public void setDormitoryLongitude(Double dormitoryLongitude) { this.dormitoryLongitude = dormitoryLongitude; }

    public Integer getGpsRadiusMeters() { return gpsRadiusMeters; }
    public void setGpsRadiusMeters(Integer gpsRadiusMeters) { this.gpsRadiusMeters = gpsRadiusMeters; }

    public Boolean getRoomPhotoValidationEnabled() { return roomPhotoValidationEnabled; }
    public void setRoomPhotoValidationEnabled(Boolean roomPhotoValidationEnabled) { this.roomPhotoValidationEnabled = roomPhotoValidationEnabled; }

    public String getApplicableDays() { return applicableDays; }
    public void setApplicableDays(String applicableDays) { this.applicableDays = applicableDays; }

    public Boolean getIsDefault() { return isDefault; }
    public void setIsDefault(Boolean isDefault) { this.isDefault = isDefault; }

    // ✅ 신규
    public Long getScheduleId() { return scheduleId; }
    public void setScheduleId(Long scheduleId) { this.scheduleId = scheduleId; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }

    public String getCreatedBy() { return createdBy; }
    public void setCreatedBy(String createdBy) { this.createdBy = createdBy; }
}